### PR TITLE
Speed up epoch summaries

### LINKED ISSUE #49: summarizer running slower than chain
I have a chaind instance that was offline for a while and caught up, except for the summarizer which is running slower than chain time. Specifically, the epoch summarizer is taking ~6-8 minutes per epoch processed. 

Digging into the trace logs, it looks like the bottleneck is the query at https://github.com/wealdtech/chaind/blob/36338be9de64f59a436459404a1700c6a86e4ece/services/chaindb/postgresql/validators.go#L302, which is taking most of those 6-8 minutes. (I re-ran the query manually and had the same result). The slowness is due to the `WHERE f_validator_index = ANY(...)` where the array has ~400k entries.

I can think of some ways to improve performance here (one would be to do a negation, and only pass the list of unwanted validators, which is tiny in comparison). 

But before doing that, I wanted to check if such a change would be welcome.  

A related and more basic question is if anyone else is running into this performance issue - seems suspicious if I'm the only one (but I don't see anything obviously amiss with my postgres instance).

### PR DESCRIPTION
**Pull request by henridf**
The epoch summarizer runs a select on the `t_validator_balances` table
with a filter like `f_validator_index = ANY(ARRAY[...])` where the array
lists all active validators (~400k at this time). This is taking a
long time to complete.

To speed things up, rather than pass this huge array in the query,
fetch all rows and do the filtering in chaind.

On my setup I'm seeing epoch summaries computed in few seconds (vs
several minutes previously).

fixes #49

### CODE REVIEW

File: services/chaindb/postgresql/validators.go [hunk 1]
```diff
@@ -279,13 +279,12 @@ func (s *Service) ValidatorsByIndex(ctx context.Context, indices []phase0.Valida
 	return validators, nil
 }
 
-// ValidatorBalancesByIndexAndEpoch fetches the validator balances for the given validators and epoch.
```
Comment (mcdee) [inline services/chaindb/postgresql/validators.go:282]: `ValidatorBalancesByIndexAndEpoch` should remain as a function, as it is public and may be used by other codebases when fetching information from the chain database.  Please retain this function, and create a new function for `ValidatorBalancesByEpoch`.

File: services/summarizer/standard/epoch.go [hunk 2]
```diff
@@ -120,34 +121,36 @@ func (s *Service) validatorSummaryStatsForEpoch(ctx context.Context,
...
+	for i, validator := range validators {
 		switch {
 		case validator.ActivationEpoch == epoch:
 			summary.ActiveValidators++
 			summary.ActivatingValidators++
-			activeIndices = append(activeIndices, validator.Index)
+			activeValidators[i] = true
 		case validator.ExitEpoch == epoch:
 			summary.ExitingValidators++
+			activeValidators[i] = false
```
Comment (mcdee) [inline services/summarizer/standard/epoch.go:141]: This value defaults to false so no need to set it explicitly here.

File: services/summarizer/standard/epoch.go [hunk 3]
```diff
@@ -120,34 +121,36 @@ func (s *Service) validatorSummaryStatsForEpoch(ctx context.Context,
...
 		case validator.ActivationEpoch <= epoch &&
 			validator.ExitEpoch > epoch:
 			summary.ActiveValidators++
-			activeIndices = append(activeIndices, validator.Index)
+			activeValidators[i] = true
 		case validator.ActivationEligibilityEpoch <= epoch &&
 			validator.ActivationEpoch != s.farFutureEpoch &&
 			validator.ActivationEpoch > epoch:
 			summary.ActivationQueueLength++
+			activeValidators[i] = false
```
Comment (mcdee) [inline services/summarizer/standard/epoch.go:150]: This value defaults to false so no need to set it explicitly here.

**Comment by henridf** [comment]
Feedback addressed in [447e1ca](https://github.com/wealdtech/chaind/pull/50/commits/447e1cad98b4d747b5a20f914d5d7006ea655dda).

File: services/chaindb/postgresql/validators.go [hunk 4]
```diff
@@ -279,6 +279,55 @@ func (s *Service) ValidatorsByIndex(ctx context.Context, indices []phase0.Valida
 	return validators, nil
 }
 
+// ValidatorBalancesByEpoch fetches the validator balances for the given validators and epoch.
```
Comment (mcdee) [inline services/chaindb/postgresql/validators.go:282]: This should be just "given epoch."

File: services/chaindb/service.go [hunk 5]
```diff
@@ -231,13 +231,12 @@ type ValidatorsProvider interface {
 	// ValidatorsByIndex fetches all validators matching the given indices.
 	ValidatorsByIndex(ctx context.Context, indices []phase0.ValidatorIndex) (map[phase0.ValidatorIndex]*Validator, error)
 
-	// ValidatorBalancesByIndexAndEpoch fetches the validator balances for the given validators and epoch.
-	ValidatorBalancesByIndexAndEpoch(
+	// ValidatorBalancesByEpoch fetches all validator balances for the given epoch.
```
[the remark this answers is not in the document]
Comment (henridf) [reply services/chaindb/service.go:234]: oops

**Pull request merged**